lib/services/plugins.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package services
+
+import (
+	"bytes"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// MarshalPlugin marshals Plugin resource to JSON.
+func MarshalPlugin(plugin types.Plugin, opts ...MarshalOption) ([]byte, error) {
+	if err := plugin.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cfg, err := CollectOptions(opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	switch plugin := plugin.(type) {
+	case *types.PluginV1:
+		if !cfg.PreserveResourceID {
+			copy := *plugin
+			copy.SetResourceID(0)
+			plugin = &copy
+		}
+		var buf bytes.Buffer
+		err := (&jsonpb.Marshaler{}).Marshal(&buf, plugin)
```

Comment:
`jsonpb` will not use the gogoproto json tags, FYI. I don't think we use it anywhere else, why here?

Reply:
`jsonpb` is used here, as `encoding/json` can not deserialize interface fields (which are introduced by `oneof`). It is used for some other types which contain `oneof` as well, e.g.:

https://github.com/gravitational/teleport/blob/799219f07a94c3133b659a9aacae24fb4a419c99/api/types/authentication.go#L735-L743

https://github.com/gravitational/teleport/blob/799219f07a94c3133b659a9aacae24fb4a419c99/api/proto/teleport/legacy/types/types.proto#L2805-L2809

(although here, `MarshalJSON` and `UnmarshalJSON` are defined instead).

gogoproto fields are leftover from before I discovered this limitation, I should probably remove those.

`encoding/json` seems to match fields case-insensitively when deserializing. https://stackoverflow.com/questions/49006073/json-unmarshal-struct-case-sensitively

Reply:
We should lowercase and snake_case the field names (like it's done for `MFADevice`) then.

Reply:
Done in 1a9bc3f157. Just to confirm, this is the result (prettified):

```
{
  "kind": "plugin",
  "version": "v1",
  "metadata": {
    "Name": "foobar"
  },
  "spec": {
    "slackAccessPlugin": {
      "fallbackChannel": "#access-requests"
    }
  },
  "status": {},
  "credentials": {
    "oauth2AccessToken": {
      "accessToken": "access_token",
      "refreshToken": "refresh_token",
      "expires": "2023-02-01T18:08:47.552420873Z"
    }
  }
}
```

`Metadata` is an embedded field, so it defines its own naming.